mri/scripts/gridsearch.py

```diff
@@ -0,0 +1,289 @@
+        if metric == 'ssim' or metric == 'psnr' or metric == 'accuracy':
+            metric_direction = True
+        elif metric == 'nrmse':
+            metric_direction = False
+        else:
+            raise ValueError('Cannot automatically find out metric direction, '
+                             'please specify metric direction')
+    # get best runs
+    if metric_direction:
+        best_metric = list_metric.max()
+        best_idx = list_metric.argmax()
+    else:
+        best_metric = list_metric.min()
+        best_idx = list_metric.argmin()
+
+    return best_metric, best_idx
+
+
+def launch_grid(kspace_data, reconstructor_class, reconstructor_kwargs,
+                fourier_op=None, linear_params=None, regularizer_params=None,
+                optimizer_params=None, fourier_params=None,
+                compare_metric_details=None, n_jobs=1, verbose=0):
+    """This function launches off reconstruction for a grid specified
+    through use of kwarg dictionaries.
+
+    Dictionary Convention
+    ---------------------
+    These dictionaries each defined to follow the convention:
+    Each dictionary has a key `init_class` that specifies the
+    initialization class for the operator (exception to
+    this is 'optimizer_params'). Later we have key `kwargs` that holds
+    all the input arguments that can be passed as a keyword dictionary.
+    Each value in this keyword dictionary ,ust be a list of all
+    values you want to search in gridsearch.
+
+    This function finds the search space of parameters and
+    sets up right parameters for '_reconstruct_case' function.
+    Please check the example code for more details.
+
+    Parameters
+    ----------
+    kspace_data: np.ndarray
+        the kspace data for reconstruction
+    reconstructor_class: class
+        reconstructor class
+    reconstructor_kwargs: dict
+        extra kwargs for reconstructor
+    fourier_op: object of class FFT
+        this defines the fourier operator. for NonCartesianFFT, please make
+        fourier_op as `None` and pass fourier_params to allow
+        parallel execution
+    linear_params: dict, default None
+        dictionary for linear operator parameters
+        if None, a sym8 wavelet is chosen
+    regularizer_params: dict, default None
+        dictionary for regularizer operator parameters
+        if None, mu=0, ie no regularization is done
+    optimizer_params: dict, default None
+        dictionary for optimizer key word arguments
+        if None, a FISTA optimization is done for 100 iterations
+    fourier_params: dict, default None
+            holds dictionary with init_class pointing to fourier
+            class to be used and args having keyword arguments for
+            initialization
+            This is passed only if fourier_op is None so that fourier_op can be
+            made on spot during reconstruction.
+            NOTE: We declare fourier operator inside this function to allow
```

Comment:
I don't think this is what we want to do. This boils down to the same situation as before. You should be able to create this dictionary yourself from the `fourier_op` passed in by the user.
Basically, you would have a test `if isinstance(fourier_op, NFFT): fourier_params = {...`.
(and you would set `fourier_op` to `None`). This way the user doesn't have to "write" the fourier operator twice.

Reply:
Ahh, that makes more sense...